runtime/src/comm/ofi/comm-ofi-oob-pmi2.c

```diff
@@ -286,6 +291,51 @@ void decode_kvs(char* raw, const char* enc, size_t size) {
 }
 
 int chpl_comm_ofi_oob_locales_on_node(void) {
-  // TODO: figure out how to implement this correctly
-  return 1;
+  int count = 0;
+  if (PMI_Get_clique_size != NULL) {
+    PMI_CHK(PMI_Get_clique_size(&count));
+    DBG_PRINTF(DBG_OOB, "PMI_Get_clique_size returned %d", count);  
+  } else {
+    // do an allgather of hostname hashes to determine the locales on the same node as us
+    // assumes each hostname has a unique hash
+
+    char hostname[256];
+    int rc = gethostname(hostname, sizeof(hostname));
+    CHK_TRUE(rc == 0);
+
+    uint64_t hash = 0;
+    for (int i = 0; i < sizeof(hostname); i++) {
+      char c = hostname[i];
+      if (c == '\0') {
+        break;
+      }
+      // The hash code is borrowed from gasnet including the comment.
+
+      /* The "c = ..." squeezes ASCII down to 6 bits, while encoding
+       * all chars valid in hostnames and IP addresses (IPV4 and IPV6).
+       * A unique value is assigned to each of the digits, the lower
+       * case letters, '-', '.' and ':'.  The upper case letters map
+       * to the same values as the corresponding lower-case.
+       */
+      c = ((c & 0x40) >> 1) | (c & 0x1f);
+      hash = ((hash << 6) | ((hash >> 58) & 0x3F)) ^ c;
+    }
+
+    // get the hashes for all locales
+
+    uint32_t *hashes = NULL;
+    CHK_SYS_CALLOC(hashes, chpl_numNodes);
+    chpl_comm_ofi_oob_allgather(&hash, hashes, sizeof(*hashes));
+
+    // count the number of hashes that match ours
+
+    for (int i = 0; i < chpl_numNodes; i++) {
+      if (hashes[i] == hash) {
+        count++;
+      }
+    }
+    sys_free(hashes);
```

Comment:
This looks a little odd to me that we allocate with a `CHK_SYS_CALLOC` wrapper, but just call the plain one here. Might be nicer for symmetry to add a `CHK_SYS_FREE` wrapper and use that.

Reply:
Will do.